controllers/machinedeployment_rolling_test.go

```diff
@@ -219,3 +219,188 @@ func TestReconcileNewMachineSet(t *testing.T) {
+					Replicas: pointer.Int32Ptr(0),
+				},
+				Status: clusterv1.MachineSetStatus{
+					AvailableReplicas: 2,
+				},
+			},
+			oldMachineSets: []*clusterv1.MachineSet{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "foo",
+						Name:      "2replicas",
+					},
+					Spec: clusterv1.MachineSetSpec{
+						Replicas: pointer.Int32Ptr(2),
+					},
+					Status: clusterv1.MachineSetStatus{
+						AvailableReplicas: 2,
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "foo",
+						Name:      "1replicas",
+					},
+					Spec: clusterv1.MachineSetSpec{
+						Replicas: pointer.Int32Ptr(1),
+					},
+					Status: clusterv1.MachineSetStatus{
+						AvailableReplicas: 1,
+					},
+				},
+			},
+			expectedOldMachineSetsReplicas: 0,
+		},
+		{
+			name: "RollingUpdate strategy: It does not scale down old MachineSets when above maxUnavailable",
+			machineDeployment: &clusterv1.MachineDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+					Name:      "bar",
+				},
+				Spec: clusterv1.MachineDeploymentSpec{
+					Strategy: &clusterv1.MachineDeploymentStrategy{
+						Type: clusterv1.RollingUpdateMachineDeploymentStrategyType,
+						RollingUpdate: &clusterv1.MachineRollingUpdateDeployment{
+							MaxUnavailable: intOrStrPtr(2),
+							MaxSurge:       intOrStrPtr(3),
+						},
+					},
+					Replicas: pointer.Int32Ptr(10),
+				},
+			},
+			newMachineSet: &clusterv1.MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+					Name:      "bar",
+				},
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: pointer.Int32Ptr(5),
+				},
+				Status: clusterv1.MachineSetStatus{
+					Replicas:          5,
+					ReadyReplicas:     0,
+					AvailableReplicas: 0,
+				},
+			},
+			oldMachineSets: []*clusterv1.MachineSet{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "foo",
+						Name:      "8replicas",
+					},
+					Spec: clusterv1.MachineSetSpec{
+						Replicas: pointer.Int32Ptr(8),
+					},
+					Status: clusterv1.MachineSetStatus{
+						Replicas:          10,
+						ReadyReplicas:     8,
+						AvailableReplicas: 8,
+					},
+				},
+			},
+			expectedOldMachineSetsReplicas: 8,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			g.Expect(clusterv1.AddToScheme(scheme.Scheme)).To(Succeed())
+
+			resources := []client.Object{
+				tc.machineDeployment,
+			}
+
+			allMachineSets := append(tc.oldMachineSets, tc.newMachineSet)
+			for key := range allMachineSets {
+				resources = append(resources, allMachineSets[key])
+			}
+
+			r := &MachineDeploymentReconciler{
+				Client:   fake.NewClientBuilder().WithObjects(resources...).Build(),
+				recorder: record.NewFakeRecorder(32),
+			}
+
+			err := r.reconcileOldMachineSets(ctx, allMachineSets, tc.oldMachineSets, tc.newMachineSet, tc.machineDeployment)
+			if tc.error != nil {
+				g.Expect(err.Error()).To(BeEquivalentTo(tc.error.Error()))
```

Comment:
if tc.error != nil and err == nil this will lead to a nil pointer, probably better if we guard against that. (same in TestReconcileNewMachineSet above)

Reply:
Fixed. Thanks!